src/unix/kqueue.c

```diff
@@ -359,7 +359,8 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
       if (ev->flags & EV_ERROR)
         revents |= POLLERR;
 
-      if ((ev->flags & EV_EOF) && (w->pevents & UV__POLLRDHUP))
+      if ((ev->filter == EVFILT_READ || ev->filter == EVFILT_EXCEPT) &&
+          (ev->flags & EV_EOF) && (w->pevents & UV__POLLRDHUP))
         revents |= UV__POLLRDHUP;
```

Comment:
Might read a little easier in "trapped" style, like below, otherwise LGTM.
```suggestion
      if (ev->filter == EVFILT_READ || ev->filter == EVFILT_EXCEPT)
        if (ev->flags & EV_EOF)
          if (w->pevents & UV__POLLRDHUP)
            revents |= UV__POLLRDHUP;
```
You can also write the last two lines as `revents |= (w->pevents & UV__POLLRDHUP);` - one less branch.